Mathlib/LinearAlgebra/Contraction.lean

```diff
@@ -0,0 +1,347 @@
+/-
+Copyright (c) 2020 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash, Antoine Labelle
+
+! This file was ported from Lean 3 source module linear_algebra.contraction
+! leanprover-community/mathlib commit 657df4339ae6ceada048c8a2980fb10e393143ec
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.LinearAlgebra.Dual
+import Mathlib.LinearAlgebra.Matrix.ToLin
+
+/-!
+# Contractions
+
+Given modules $M, N$ over a commutative ring $R$, this file defines the natural linear maps:
+$M^* \otimes M \to R$, $M \otimes M^* \to R$, and $M^* \otimes N → Hom(M, N)$, as well as proving
+some basic properties of these maps.
+
+## Tags
+
+contraction, dual module, tensor product
+-/
+
+-- Porting note: universe metavariables behave oddly
+universe w u v₁ v₂ v₃ v₄
+
+variable {ι : Type w} (R : Type u) (M : Type v₁) (N : Type v₂)
+  (P : Type v₃) (Q : Type v₄)
+
+-- Porting note: this does not appear to work; see below
+attribute [local ext] TensorProduct.ext
```

Comment:
`local ext high` works here.

Reply:
Thanks for figuring that out. I updated the porting note here and removed the manual ext lemma calls.